### DISCUSSION TITLE
Blueprint before_app_first_request vs record_once usage discrepancy

### DISCUSSION THREAD

**Discussion by TheDavidFactor** [pallets/flask#5166] [Q&A] [answered]
I'm working on upgrading from Flask 2.2.x to 2.3.x and I use the .before_app_first_request decorator in one of my blueprints. Following the deprecation warning, I replaced .before_app_first_request with .record_once and it was not a drop-in replacement. Not sure if this an issue in code implementation or unclear documentation.

```
from flask import Blueprint, Flask, current_app

blueprint = Blueprint('my_blueprint', __name__)
@blueprint.before_app_first_request
def my_test_function():
    current_app.logger.info('testing')

app = Flask(__name__)
app.register_blueprint(blueprint)
app.run()
```

replace .before_app_first_request with .record_once and it will no longer run

Environment:

- Python version: 3.10.11
- Flask version: 2.2.2

**Comment by davidism** [comment, marked as answer]
See the documentation for `record_once` and `record`: https://flask.palletsprojects.com/en/2.3.x/api/#flask.Blueprint.record. The decorated function takes a `state` argument.

```python
@blueprint.record_once
def my_test_function(state):
    state.app.logger.info("testing")
```

**Comment by TheDavidFactor** [comment]
Thank you for the very fast response. I didn't read through the documentation far enough.
